### PR TITLE
Python: Improve Avro Parse speed

### PR DESCRIPTION
**Pull request by rustyconover**
Cache the hash value of the Struct types so that when new Record objects are created time is not spent recomputing the hash value.

Change the implementation of the Record class to cache the mapping from field name to position.  Also utilize an array rather than a dictionary for these lookups.

### CODE REVIEW

**Review by Fokko — COMMENTED** [review]
One suggest, looks great, and again a major speedup!

Current master:
```
➜  iceberg-2 git:(master) hyperfine --warmup 1 "python3 /tmp/vo.py"
Benchmark 1: python3 /tmp/vo.py
  Time (mean ± σ):      2.654 s ±  0.027 s    [User: 2.790 s, System: 1.964 s]
  Range (min … max):    2.626 s …  2.708 s    10 runs
```

The branch:
```
➜  iceberg-2 git:(fix_python_record_speed) hyperfine --warmup 1 "python3 /tmp/vo.py"        
Benchmark 1: python3 /tmp/vo.py
  Time (mean ± σ):      2.001 s ±  0.049 s    [User: 2.175 s, System: 1.883 s]
  Range (min … max):    1.959 s …  2.132 s    10 runs
```

And with the `lru_cache`:
```
➜  iceberg-2 git:(fix_python_record_speed) hyperfine --warmup 1 "python3 /tmp/vo.py"
Benchmark 1: python3 /tmp/vo.py
  Time (mean ± σ):      1.997 s ±  0.032 s    [User: 2.128 s, System: 1.981 s]
  Range (min … max):    1.966 s …  2.070 s    10 runs
```

Thanks for working on this!

File: python/pyiceberg/typedef.py [hunk 1]
```diff
@@ -127,18 +127,23 @@ def json(self, exclude_none: bool = True, exclude: Optional[Set[str]] = None, by
         )
 
 
+CACHED_STRUCT_TO_FIELD_NAME_POSITIONS: Dict[StructType, List[str]] = {}
```
Comment (Fokko) [inline python/pyiceberg/typedef.py:130]:
```suggestion
@lru_cache
def _get_struct_fields(struct_type: StructType) -> List[str]:
    return [field.name for field in struct_type.fields]
```

File: python/pyiceberg/types.py [hunk 2]
```diff
@@ -314,6 +316,10 @@ def __getnewargs__(self) -> Tuple[NestedField, ...]:
         """A magic function for pickling the StructType class."""
         return self.fields
 
+    def __hash__(self) -> int:
```
Comment (rdblue) [inline python/pyiceberg/types.py:319]: Is there a way to automatically memoize this instead of using a field?

Comment (rustyconover) [reply python/pyiceberg/types.py:319]: Not that I've determined, as python presumes the hash value can change as the instance is mutated.

File: python/pyiceberg/typedef.py (hunk 1 above)
Comment (rustyconover) [reply python/pyiceberg/typedef.py]: Changed in latest commit.

**Comment by Fokko** [comment]
Thanks for fixing this @rustyconover and @rdblue for the review!

**Pull request merged**